### PR TITLE
ARROW-4615: [C++] Add checked_pointer_cast

### PR DESCRIPTION
**Pull request by xhochy**

### CODE REVIEW

File: cpp/src/arrow/array-binary-test.cc [hunk 1]
```diff
@@ -184,7 +185,7 @@ TEST_F(TestStringArray, TestSliceGetString) {
   std::shared_ptr<Array> array;
   ASSERT_OK(builder.Finish(&array));
   auto s = array->Slice(1, 10);
-  auto arr = std::dynamic_pointer_cast<StringArray>(s);
+  auto arr = checked_pointer_cast<StringArray>(s);
```
Comment (pitrou) [inline cpp/src/arrow/array-binary-test.cc]: In tests, I think we would like to keep `dynamic_pointer_cast` (e.g. if tests are built in release mode)

**Review by pitrou — COMMENTED** [review]
Just one comment. I think it's valuable to keep dynamic casts in tests.

File: cpp/src/arrow/array-binary-test.cc (hunk 1 above)
Comment (xhochy) [reply cpp/src/arrow/array-binary-test.cc]: This patch came out of the Fedora compiler problems, reverted it.

Comment (pitrou) [reply cpp/src/arrow/array-binary-test.cc]: I don't think we should aim at making Arrow work without RTTI.

**Comment by pitrou** [comment]
Seems there are various CI failures.

**Comment by xhochy** [comment]
@pitrou CI failure are now fixed.

**Review by pitrou — APPROVED** [review]
LGTM

**Pull request closed without merging**